[WIP] Add Linux kernel profiling lab (#282)
linux-kernel-labs/linux: Documentation/teaching/labs/kernel_profiling.rst

================
Kernel Profiling
================

Lab objectives
==============

  * Familiarize yourself with the basics of Linux kernel profiling
  * Understanding basic profiling tools
  * Learning profiling methodologies and good practices

Overview
========

Up until now we have studied how the different components of the Linux kernel
work, and how to write drivers that interface with them in order to provide

> support for devices or protocols.This has helped us understand how the Linux

Reviewer: ```suggestion
support for devices or protocols. This has helped us understand how the Linux
```

kernel works, but most people will not get to write kernel drivers.

Nonetheless, the skills learned will help us to write applications that better
integrate with the whole operating system. In order to do this, one has to have
a good view of both the user space and the kernel space.

> This session will try to merge the work we have done up until now in the kernel

Reviewer: ```suggestion
This session aims to merge the work we have done up until now in the kernel
```

space with real world use cases where we do not write kernel space code, but we
look through the kernel using profiling tools, in order to debug issues that

> we're having when writing regular, low level, applications.

Reviewer: ```suggestion
we're having when writing regular, low-level, applications.
```

> Another focus of this laboratory will be learning a general methodology for

Reviewer: ```suggestion
Another focus of this session will be learning a general methodology for
```

debugging software issues, and we will approach some tools that give us insight
from the kernel on the way our application runs.

> Profiling tools

Reviewer: ```suggestion
Profiling Tools
```

===============

The main tool that we will focus our attention on is ``perf``, which offers
support for tracing applications, and also inspecting general aspects of the
system. We will also be using debugging tools that most people have used in
their day to day life, such as ``htop``, ``ps``, ``lsof`` and others.

perf
----

> ``perf``,  is a tool that instruments the CPU using

Reviewer: ```suggestion
``perf``  is a tool that instruments the CPU using
```

       kallsyms        Searches running kernel for symbols
       kmem            Tool to trace/measure kernel memory properties
       kvm             Tool to trace/measure kvm guest os
       list            List all symbolic event types
       lock            Analyze lock events
       mem             Profile memory accesses
       record          Run a command and record its profile into perf.data
       report          Read perf.data (created by perf record) and display the profile
       sched           Tool to trace/measure scheduler properties (latencies)
       script          Read perf.data (created by perf record) and display trace output
       stat            Run a command and gather performance counter statistics
       test            Runs sanity tests.
       timechart       Tool to visualize total system behavior during a workload
       top             System profiling tool.
       version         display the version of perf binary
       probe           Define new dynamic tracepoints

     See 'perf help COMMAND' for more information on a specific command.

In the output above we can see all of perf's subcommands together with a

> description of their functionality the most significant of which are:

Reviewer: ```suggestion
description of their functionality, the most significant of which are:
```

offer a higher level way of looking at events or data, without needing to know
the intricacies of the command. An example of this is the ``iosnoop.sh`` script,
which displays what I/O transfers are taking place.

ps
--

``ps`` is the linux tool that allows us to monitor the processes that are
running at a given time on the machine, including the kernel threads. This is a
simple and easy to use way of checking at a glance what processes are running on
the CPU, and what is their CPU and memory usage.

In order to list all the processes running, we use to ``ps aux`` command in the
following way:

.. code-block:: c

   TODO

One information of note is that the 7th column represents the that of the
process, ``S`` meaning suspended, ``D`` suspended due to I/O, and ``R`` meaning
running.

time
----

The ``time`` command allows us to inspect the amount of time spent by a
process in I/O, running the application code, or running code in kernel space.
This can be useful in order to find out whether an application's issue comes
from running too much in kernel space, so it has some overhead when it does
system calls, or the issue is in the user code.

.. code-block:: c

   TODO

top
---

> ``top`` is an application that is found on most system which lists in real time

Reviewer: ```suggestion
``top`` is an application that is found on most systems which lists in real time
```

the applications that are running on the system. ``top`` runs interactively, and
it auto-refreshes its output, as opposed to ``ps``. We use this tool when we
want a high level of continuous monitoring.

> Profiling methodology

Reviewer: ```suggestion
Profiling Methodology
```

=====================

> When doing profiling, our goal is to research the cause of a problem. Usually

Reviewer: ```suggestion
When doing profiling, our goal is to identify the cause of a problem. Usually
```

this problem is observed by someone when their application doesn't work as
expected. When we say that an application did not work as expected, this can
mean different things for different people. For example, one person might
complain that the application has a slowdown, while another might say that the
application runs on the CPU, but it doesn't output anything.

The first step in any problem solving context is to understand the default
behaviour of the application we're trying to debug, and to make sure that it is
now not running in the expected parameters.

> Profiling memory usage

Reviewer: ```suggestion
Profiling Memory Usage
```

======================

> Profiling I/O problems

Reviewer: ```suggestion
Profiling I/O Problems
```

    io-app       889    WS   254,0    4800512      1310720     2.60
    io-app       889    WS   254,0    4803072      1310720     2.58
    [...]

From this output we see that the ``io-app`` is reading in a loop from the fact
that the first block ``4800512`` is repeating, and that it is doing big reads,
since it is reading one megabyte fer request. This constant looping adds the
load to the system that we're experiencing.

Profiling scheduling issues
===========================

For scheduling learning how to debug scheduling issues, we'll be taking a look

Exercises
=========

.. include:: ../labs/exercises-summary.hrst
.. |LAB_NAME| replace:: kernel_profiling

1. Investigating reduced responsiveness
---------------------------------------

The ``io.ko`` module, located in the ``kernel_profiling/1-io`` directory,
decreases the system's responsiveness when inserted. We see that the command
line stutters when typing commands, but when running top, we see that the
system's load is not high, and there aren't any processes that are hogging
resources.

Find out what the ``io.ko`` module is doing and why is it leading to the
stuttering effect that we experience.

> .. hint:: Trace all the functions being called and check where the CPU is

Reviewer: ```suggestion
.. hint::

    Trace all the functions being called and check where the CPU is
```

    manager, we will be downloading the ``perf`` binary from `this
    <http://swarm.cs.pub.ro/~sweisz/perf>`_ link. Download the application in
    the ``skels/kernel_profiling`` directory, and grant in execution
    permissions.

.. warning::

    When running ``perf``, make sure that you're running the downloaded version,
    not the version in the ``PATH`` variable.

.. note::

    When going through this session's exercises, we will have to run command in
    parallel. In order to do this, we will have to connect to the virtual machine
    using SSH. We recommend using the ``core-image-sato-sdk-qemu`` image, since it
    has the tools that we need. To run the virtual machine using the
    ``core-image-sato-sdk-qemu`` file system, uncomment line 16 in the
    ``qemu/Makefile`` file.

.. note::

    If you wish to run the ``perf-tools`` based scripts that we have included in
    the repository, such as ``iosnoop.sh``, you will have to grant it execution
    privilleges, in order to be copied to the virtual machine file system.

.. note::

    In order to improve the course of SO2, its components and the way it is
    conducted, your opinions are very useful to us.

>     `curs.upb.ro platform <https://curs.upb.ro/mod/feedbackadm/view.php?id=26596>`_

Reviewer: Please update link to: https://curs.upb.ro/2021/mod/feedbackadm/view.php?id=16907

>     The form is anonymous and is active between May 24 and June 4, 2021. The

Reviewer: ```suggestion
    The form is anonymous and is active between May 21 and June 2, 2022. The
```
